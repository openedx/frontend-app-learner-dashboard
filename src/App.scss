// frontend-app-*/src/index.scss
@use "@openedx/paragon/styles/css/core/custom-media-breakpoints" as paragonCustomMediaBreakpoints;

$fa-font-path: "~font-awesome/fonts";
@import "~font-awesome/scss/font-awesome";

$input-focus-box-shadow: var(--pgn-elevation-form-input-base); // hack to get upgrade to paragon 4.0.0 to work

@import "~@edx/frontend-component-header/dist/index";
@import "~@edx/frontend-component-footer/dist/_footer";

.text-ellipsis {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.alert.alert-info .alert-icon {
  color: black;
}

#root {
  // Removing a odd 1.5 scaling on checkboxes.:
  input[type=checkbox] {
    transform: none;
  }
  display: flex;
  flex-direction: column;
  min-height: 100vh;

  main {
    flex-grow: 1;
  }

  header {
    flex: 0 0 auto;

    .logo {
      display: block;
      box-sizing: content-box;
      position: relative;
      top: 0.1em;
      height: 1.75rem;
      margin-right: 1rem;
      img {
        display: block;
        height: 100%;
      }
    }
  }

  footer {
    flex: 0;
  }
}

#paragon-portal-root {
  .pgn__modal-layer {
    .pgn__modal-close-container {
      right: 1rem !important;
    }
  }
  .confirm-modal .pgn__modal-body {
    overflow: hidden;
  }
}
