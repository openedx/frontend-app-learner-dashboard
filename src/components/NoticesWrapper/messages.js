import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  error404Message: {
    id: 'learner-dash.notices.error404Message',
    defaultMessage: 'This probably happened because the notices plugin is not installed on platform.',
    description: 'Error message when notices API returns 404',
  },
});

export default messages;
