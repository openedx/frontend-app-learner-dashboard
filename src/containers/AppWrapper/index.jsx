import PropTypes from 'prop-types';

export const AppWrapper = ({
  children,
}) => children;
AppWrapper.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]).isRequired,
};

export default AppWrapper;
