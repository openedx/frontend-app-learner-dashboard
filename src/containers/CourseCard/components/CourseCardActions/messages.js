import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  beginCourse: {
    id: 'learner-dash.courseCard.actions.beginCourse',
    description: 'Course card begin-course button text',
    defaultMessage: 'Begin Course',
  },
  resume: {
    id: 'learner-dash.courseCard.actions.resume',
    description: 'Course card resume button text',
    defaultMessage: 'Resume',
  },
  viewCourse: {
    id: 'learner-dash.courseCard.actions.viewCourse',
    description: 'Course card view-course button text',
    defaultMessage: 'View Course',
  },
  selectSession: {
    id: 'learner-dash.courseCard.actions.selectSession',
    description: 'Course card select-session button text',
    defaultMessage: 'Select Session',
  },
});

export default messages;
