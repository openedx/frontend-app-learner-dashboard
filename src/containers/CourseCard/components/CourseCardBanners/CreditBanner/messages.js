import { defineMessages } from '@openedx/frontend-base';

const messages = defineMessages({
  error: {
    id: 'learner-dash.courseCard.banners.credit.error',
    description: 'Error message for credit transaction with support email link',
    defaultMessage: 'An error occurred with this transaction.  For help, contact {supportEmailLink}.',
  },
  errorNoEmail: {
    id: 'learner-dash.courseCard.banners.credit.errorNoEmail',
    description: 'Error message for credit transaction without support email',
    defaultMessage: 'An error occurred with this transaction.',
  },
});

export default messages;
