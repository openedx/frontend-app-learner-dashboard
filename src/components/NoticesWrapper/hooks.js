import React from 'react';
import { getConfig } from '@edx/frontend-platform';
import { useIntl } from 'react-intl';

import { StrictDict } from 'utils';
import { getNotices } from './api';
import * as module from './hooks';
import messages from './messages';

/**
 * This component uses the platform-plugin-notices plugin to function.
 * If the user has an unacknowledged notice, they will be rerouted off
 * course home and onto a full-screen notice page. If the plugin is not
 * installed, or there are no notices, we just passthrough this component.
 */
export const state = StrictDict({
  isRedirected: (val) => React.useState(val), // eslint-disable-line
});

export const useNoticesWrapperData = () => {
  const [isRedirected, setIsRedirected] = module.state.isRedirected();
  const { formatMessage } = useIntl();

  React.useEffect(() => {
    if (getConfig().ENABLE_NOTICES) {
      getNotices({
        onLoad: (data) => {
          if (data?.data?.results?.length > 0) {
            setIsRedirected(true);
            window.location.replace(`${data.data.results[0]}?next=${window.location.href}`);
          }
        },
        notFoundMessage: formatMessage(messages.error404Message),
      });
    }
  }, [setIsRedirected, formatMessage]);
  return { isRedirected };
};

export default useNoticesWrapperData;
