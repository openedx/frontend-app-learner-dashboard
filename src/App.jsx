import React from 'react';
import { Helmet } from 'react-helmet';

import { useIntl } from '@edx/frontend-platform/i18n';
import { logError } from '@edx/frontend-platform/logging';
import { initializeHotjar } from '@edx/frontend-enterprise-hotjar';

import { ErrorPage, AppContext } from '@edx/frontend-platform/react';
import FooterSlot from '@openedx/frontend-slot-footer';
import { Alert } from '@openedx/paragon';

import { RequestKeys } from 'data/constants/requests';
import store from 'data/store';
import {
  selectors,
  actions,
} from 'data/redux';
import { reduxHooks } from 'hooks';
import Dashboard from 'containers/Dashboard';

import track from 'tracking';

import fakeData from 'data/services/lms/fakeData/courses';

import AppWrapper from 'containers/WidgetContainers/AppWrapper';
import LearnerDashboardHeader from 'containers/LearnerDashboardHeader';

import { getConfig } from '@edx/frontend-platform';
import messages from './messages';
import './App.scss';

export const App = () => {
  const { authenticatedUser } = React.useContext(AppContext);
  const { formatMessage } = useIntl();
  const isFailed = {
    initialize: reduxHooks.useRequestIsFailed(RequestKeys.initialize),
    refreshList: reduxHooks.useRequestIsFailed(RequestKeys.refreshList),
  };
  const hasNetworkFailure = isFailed.initialize || isFailed.refreshList;
  const { supportEmail } = reduxHooks.usePlatformSettingsData();
  const loadData = reduxHooks.useLoadData();

  React.useEffect(() => {
    if (authenticatedUser?.administrator || getConfig().NODE_ENV === 'development') {
      window.loadEmptyData = () => {
        loadData({ ...fakeData.globalData, courses: [] });
      };
      window.loadMockData = () => {
        loadData({
          ...fakeData.globalData,
          courses: [
            ...fakeData.courseRunData,
            ...fakeData.entitlementData,
          ],
        });
      };
      window.store = store;
      window.selectors = selectors;
      window.actions = actions;
      window.track = track;
    }
    if (getConfig().HOTJAR_APP_ID) {
      try {
        initializeHotjar({
          hotjarId: getConfig().HOTJAR_APP_ID,
          hotjarVersion: getConfig().HOTJAR_VERSION,
          hotjarDebug: !!getConfig().HOTJAR_DEBUG,
        });
      } catch (error) {
        logError(error);
      }
    }
  }, [authenticatedUser, loadData]);
  return (
    <>
      <Helmet>
        <title>{formatMessage(messages.pageTitle)}</title>
        <link rel="shortcut icon" href={getConfig().FAVICON_URL} type="image/x-icon" />
      </Helmet>
      <div>
        <AppWrapper>
          <LearnerDashboardHeader />
          <main id="main">
            {hasNetworkFailure
              ? (
                <Alert variant="danger">
                  <ErrorPage message={formatMessage(messages.errorMessage, { supportEmail })} />
                </Alert>
              ) : (
                <Dashboard />
              )}
          </main>
        </AppWrapper>
        <FooterSlot />
      </div>
    </>
  );
};

export default App;
