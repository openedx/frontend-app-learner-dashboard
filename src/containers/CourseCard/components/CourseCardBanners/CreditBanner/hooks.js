import { StrictDict } from '../../../../../utils';
import { reduxHooks } from '../../../../../hooks';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = reduxHooks.useCardCreditData(cardId);
  const { supportEmail } = reduxHooks.usePlatformSettingsData();
  if (!credit.isEligible) {
    return null;
  }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
