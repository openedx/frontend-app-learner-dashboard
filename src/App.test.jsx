import React from 'react';
import { Helmet } from 'react-helmet';
import { shallow } from '@edx/react-unit-test-utils';

import { useIntl } from '@edx/frontend-platform/i18n';
import { getConfig } from '@edx/frontend-platform';

import { RequestKeys } from 'data/constants/requests';
import { reduxHooks } from 'hooks';
import Dashboard from 'containers/Dashboard';
import LearnerDashboardHeader from 'containers/LearnerDashboardHeader';
import AppWrapper from 'containers/WidgetContainers/AppWrapper';
import { App } from './App';
import messages from './messages';

jest.mock('@edx/frontend-component-footer', () => ({ FooterSlot: 'FooterSlot' }));

jest.mock('containers/Dashboard', () => 'Dashboard');
jest.mock('containers/LearnerDashboardHeader', () => 'LearnerDashboardHeader');
jest.mock('containers/WidgetContainers/AppWrapper', () => 'AppWrapper');
jest.mock('data/redux', () => ({
  selectors: 'redux.selectors',
  actions: 'redux.actions',
  thunkActions: 'redux.thunkActions',
}));
jest.mock('hooks', () => ({
  reduxHooks: {
    useRequestIsFailed: jest.fn(),
    usePlatformSettingsData: jest.fn(),
    useLoadData: jest.fn(),
  },
}));
jest.mock('data/store', () => 'data/store');

jest.mock('@edx/frontend-platform', () => ({
  getConfig: jest.fn(() => ({})),
}));

const loadData = jest.fn();
reduxHooks.useLoadData.mockReturnValue(loadData);

let el;

const supportEmail = 'test-support-url';
reduxHooks.usePlatformSettingsData.mockReturnValue({ supportEmail });

describe('App router component', () => {
  const { formatMessage } = useIntl();
  describe('component', () => {
    const runBasicTests = () => {
      test('snapshot', () => { expect(el.snapshot).toMatchSnapshot(); });
      it('displays title in helmet component', () => {
        const control = el.instance
          .findByType(Helmet)[0]
          .findByType('title')[0];
        expect(control.children[0].el).toEqual(formatMessage(messages.pageTitle));
      });
      it('displays learner dashboard header', () => {
        expect(el.instance.findByType(LearnerDashboardHeader).length).toEqual(1);
      });
      it('wraps the header and main components in an AppWrapper widget container', () => {
        const container = el.instance.findByType(AppWrapper)[0];
        expect(container.children[0].type).toEqual('LearnerDashboardHeader');
        expect(container.children[1].type).toEqual('main');
      });
    };
    describe('no network failure', () => {
      beforeAll(() => {
        reduxHooks.useRequestIsFailed.mockReturnValue(false);
        getConfig.mockReturnValue({});
        el = shallow(<App />);
      });
      runBasicTests();
      it('loads dashboard', () => {
        const main = el.instance.findByType('main')[0];
        expect(main.children.length).toEqual(1);
        const dashboard = main.children[0];
        expect(dashboard.type).toEqual('Dashboard');
        expect(
          dashboard.matches(shallow(<Dashboard />)),
        ).toEqual(true);
      });
    });
    describe('no network failure with optimizely url', () => {
      beforeAll(() => {
        reduxHooks.useRequestIsFailed.mockReturnValue(false);
        getConfig.mockReturnValue({ OPTIMIZELY_URL: 'fake.url' });
        el = shallow(<App />);
      });
      runBasicTests();
      it('loads dashboard', () => {
        const main = el.instance.findByType('main')[0];
        expect(main.children.length).toEqual(1);
        const dashboard = main.children[0];
        expect(dashboard.type).toEqual('Dashboard');
        expect(
          dashboard.matches(shallow(<Dashboard />)),
        ).toEqual(true);
      });
    });
    describe('no network failure with optimizely project id', () => {
      beforeAll(() => {
        reduxHooks.useRequestIsFailed.mockReturnValue(false);
        getConfig.mockReturnValue({ OPTIMIZELY_PROJECT_ID: 'fakeId' });
        el = shallow(<App />);
      });
      runBasicTests();
      it('loads dashboard', () => {
        const main = el.instance.findByType('main')[0];
        expect(main.children.length).toEqual(1);
        const dashboard = main.children[0];
        expect(dashboard.type).toEqual('Dashboard');
        expect(
          dashboard.matches(shallow(<Dashboard />)),
        ).toEqual(true);
      });
    });
    describe('initialize failure', () => {
      beforeAll(() => {
        reduxHooks.useRequestIsFailed.mockImplementation((key) => key === RequestKeys.initialize);
        getConfig.mockReturnValue({});
        el = shallow(<App />);
      });
      runBasicTests();
      it('loads error page', () => {
        const main = el.instance.findByType('main')[0];
        expect(main.children.length).toEqual(1);
        const alert = main.children[0];
        expect(alert.type).toEqual('Alert');
        expect(alert.children.length).toEqual(1);
        const errorPage = alert.children[0];
        expect(errorPage.type).toEqual('ErrorPage');
        expect(errorPage.props.message).toEqual(formatMessage(messages.errorMessage, { supportEmail }));
      });
    });
    describe('refresh failure', () => {
      beforeAll(() => {
        reduxHooks.useRequestIsFailed.mockImplementation((key) => key === RequestKeys.refreshList);
        getConfig.mockReturnValue({});
        el = shallow(<App />);
      });
      runBasicTests();
      it('loads error page', () => {
        const main = el.instance.findByType('main')[0];
        expect(main.children.length).toEqual(1);
        const alert = main.children[0];
        expect(alert.type).toEqual('Alert');
        expect(alert.children.length).toEqual(1);
        const errorPage = alert.children[0];
        expect(errorPage.type).toEqual('ErrorPage');
        expect(errorPage.props.message).toEqual(formatMessage(messages.errorMessage, { supportEmail }));
      });
    });
  });
});
