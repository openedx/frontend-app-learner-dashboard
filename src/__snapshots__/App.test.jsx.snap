// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App router component component initialize failure snapshot 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Learner Home
    </title>
    <link
      rel="shortcut icon"
      type="image/x-icon"
    />
  </HelmetWrapper>
  <div>
    <AppWrapper>
      <LearnerDashboardHeader />
      <main
        id="main"
      >
        <Alert
          variant="danger"
        >
          <ErrorPage
            message="If you experience repeated failures, please email support at test-support-url"
          />
        </Alert>
      </main>
    </AppWrapper>
    <FooterSlot />
  </div>
</Fragment>
`;

exports[`App router component component no network failure snapshot 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Learner Home
    </title>
    <link
      rel="shortcut icon"
      type="image/x-icon"
    />
  </HelmetWrapper>
  <div>
    <AppWrapper>
      <LearnerDashboardHeader />
      <main
        id="main"
      >
        <Dashboard />
      </main>
    </AppWrapper>
    <FooterSlot />
  </div>
</Fragment>
`;

exports[`App router component component no network failure with optimizely project id snapshot 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Learner Home
    </title>
    <link
      rel="shortcut icon"
      type="image/x-icon"
    />
  </HelmetWrapper>
  <div>
    <AppWrapper>
      <LearnerDashboardHeader />
      <main
        id="main"
      >
        <Dashboard />
      </main>
    </AppWrapper>
    <FooterSlot />
  </div>
</Fragment>
`;

exports[`App router component component no network failure with optimizely url snapshot 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Learner Home
    </title>
    <link
      rel="shortcut icon"
      type="image/x-icon"
    />
  </HelmetWrapper>
  <div>
    <AppWrapper>
      <LearnerDashboardHeader />
      <main
        id="main"
      >
        <Dashboard />
      </main>
    </AppWrapper>
    <FooterSlot />
  </div>
</Fragment>
`;

exports[`App router component component refresh failure snapshot 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Learner Home
    </title>
    <link
      rel="shortcut icon"
      type="image/x-icon"
    />
  </HelmetWrapper>
  <div>
    <AppWrapper>
      <LearnerDashboardHeader />
      <main
        id="main"
      >
        <Alert
          variant="danger"
        >
          <ErrorPage
            message="If you experience repeated failures, please email support at test-support-url"
          />
        </Alert>
      </main>
    </AppWrapper>
    <FooterSlot />
  </div>
</Fragment>
`;
